DRAFT: Convert CircleCI build and test into GitHub Actions

wreiske opened the pull request:
## 📦 What’s Changed

This PR replaces our existing CircleCI/TravisCI pipeline with a GitHub Actions workflow (`.github/workflows/build-and-test.yml`) that:

- Reproduces the following reusable steps as individual Action steps:
  - Environment setup (core-dump dir, locale, ulimits)
  - Continuous memory-usage logging  
  - Saving the Node binary on failure  
- Leverages `actions/cache` to persist:
  - `dev_bundle` build artifacts  
  - NPM dependencies (split into two caches to avoid “MetadataTooLarge”)  
  - Computed test-groups data  
- Uploads artifacts for:
  - Core dumps  
  - Memory logs  
  - JUnit test results  
  - Full workspace snapshot  
- Orchestrates jobs with `needs:` so that:
  1. **build-env** (formerly “Get Ready”) runs first  
  2. **isolated-tests** and **test-groups** run in parallel after build-env  
  3. **docs** (JSDoc tests) runs independently  
  4. **clean-up** (balancing test-groups & saving caches) runs last  

## 🔍 Why

- **Single platform**: Consolidates CI into GitHub Actions for simpler maintenance and better integration with PR checks.  
- **Parallelization**: Leverages a matrix job for the 12 test groups to reduce duplication.  
- **Familiar tools**: Uses built-in Actions (`checkout`, `cache`, `upload-artifact`) instead of custom scripts.  
- **Visibility**: Artifacts (logs, core dumps, results) are easily accessible in the Actions UI.

## ✅ How to Test

1. 💾 **Push** this branch to GitHub.  
2. 🏗️ **Open** a draft PR against `main` (or your default branch).  
3. 🤖 **Watch** the new “Build and Test” workflow trigger under **Actions**.  
4. 🔍 **Verify**:
   - All jobs start
   - Caches are restored and saved without errors. 
   - Memory-logging runs in the background. 
   - Artifacts (core dumps, memuse.txt, JUnit XML) appear under each job. 
   - The matrix of test-groups completes all 12 partitions. 

## 📝 Notes & Future Work

- We may swap in a community JUnit reporter (e.g. `dorny/test-reporter`) for prettier test-result annotations.  
- Timeouts (`timeout-minutes`) mirror existing `no_output_timeout`, but can be tuned per job.  
- If you run into cache-key collisions, adjust `hashFiles(...)` patterns or add branch-specific prefixes.

---

_Feedback welcome!_ 🎉

wreiske committed:
Convert CircleCI build and test into GitHub Actions

Update build-and-test.yaml

wreiske committed:
Update build-and-test.yaml

wreiske committed:
Update build-and-test.yaml

wreiske commented:
Woo! Finally got tests running. That was a PITA.... Going to bed now. 💤 hopefully they finish!

wreiske committed:
Update build-and-test.yaml

wreiske commented:
Looks like several tests passed! 

The ones that are failing all seem to have this in common: 
```
Error: Depend on unknown package ddp-client?
  2|     at IsopackCache._ensurePackageLoaded (C:\tools\isobuild\isopack-cache.js:216:13)
  2|     at IsopackCache.buildLocalPackages (C:\tools\isobuild\isopack-cache.js:70:20)
  2|     at C:\tools\tool-env\isopackets.js:197:52
  2|     at Object.enterJob (C:\tools\utils\buildmessage.js:387:18)
  2|     at C:\tools\tool-env\isopackets.js:193:28
  2|     at processTicksAndRejections (node:internal/process/task_queues:105:5)
  2|     at Object.capture (C:\tools\utils\buildmessage.js:282:5)
  2|     at Console.withProgressDisplayVisible (C:\tools\console\console.js:639:14)
  2|     at Object.ensureIsopacketsLoadable (C:\tools\tool-env\isopackets.js:161:18)
  2|     at C:\tools\cli\main.js:870:3
```

wreiske committed:
Merge branch 'devel' into patch-2

nachocodoner commented:
Thank you, we’ll definitely want to merge this. We planned to migrate to GitHub Actions, covering TravisCI and CircleCI, so this progress is great. The emojis look nice.

Is there a way to limit group entries in the PR? It may be fine for others, but it feels like too much scrolling when entries only show the group number. We could show details only when a test fails clicking the parent check as we were doing so far.

What do you think?

wreiske committed:
ci: group all Meteor test jobs under a single matrix job

- Combine isolated and group test jobs into a single 'meteor-tests' matrix job
- Keeps all test runs under one expandable job in GitHub Actions UI
- Reduces workflow clutter while preserving parallel execution
- Removes old 'isolated-tests' and 'test-groups' jobs

wreiske commented:
Hmm. I agree it's noisy with all the test groups showing in the github action check log. Though it might be a good thing, since you can easily see which test groups are failing right on the PR check without having to click further in on a seperate screen. I tried grouping them under a "meteor tests", but they are still showing in the UI when failing. May have to spend more time and research on the GitHub actions side to make this less noisy. 

Any thoughts on why ddp-client is failing for these tests?

italojs commented:
> Hmm. I agree it's noisy with all the test groups showing in the github action check log. Though it might be a good thing, since you can easily see which test groups are failing right on the PR check without having to click further in on a seperate screen. I tried grouping them under a "meteor tests", but they are still showing in the UI when failing. May have to spend more time and research on the GitHub actions side to make this less noisy.
> 
> Any thoughts on why ddp-client is failing for these tests?

I agree with @wreiske's approach; it's verbose but not excessively so. I feel it's easier for a developer to visualize which group is failing without opening the action itself.

italojs commented:
Any updates about this PR @wreiske ?
we are really excited with this work

Grubba27 commented:
We migrated to GitHub Actions, thanks a lot for the pr!
